hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcServer.java

```diff
@@ -413,6 +415,19 @@ private void initSSL(ChannelPipeline p, boolean supportPlaintext)
       sslHandler.setWrapDataSize(
         conf.getInt(HBASE_SERVER_NETTY_TLS_WRAP_SIZE, DEFAULT_HBASE_SERVER_NETTY_TLS_WRAP_SIZE));
 
+      sslHandler.handshakeFuture().addListener(future -> {
+        try {
+          Certificate[] certificates = sslHandler.engine().getSession().getPeerCertificates();
+          if (certificates.length > 0) {
+            conn.clientCertificate = (X509Certificate) certificates[0];
+          } else {
```

Comment:
We should inspect the `sslHandler.engine().getNeedClientAuth()` and throw an exception if no certificate is found here and/or in the exception below. See https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java#L418-L463 for how zookeeper handles hits.

Another thing I notice in zookeeper is that they set the entire Certificate[] onto the connection, so that someone can inspect the whole certificate chain if they want. I think we should probably do that in order to avoid having to break this interface in the future if someone wants more than the head of the chain.

Reply:
I've changed the interface to use the certificate chain. I'm also logging when `getNeedClientAuth()` is true but no certificate is found. I don't think throwing an exception in this listener is any more useful than logging, it wouldn't block a connection from forming. I also don't necessarily think that it's the job of this code to verify that connections that require mutual auth are doing it. Netty is already enforcing that.